examples/gnrc_border_router: add example for WiFi ⇆ esp_now border router (#13545)
RIOT-OS/RIOT: examples/gnrc_border_router/README.md

The script `start_network.sh` enables a *ready-to-use* BR in only one command.

## Uplink

The border router will route packets between a 6Lo network (PAN) and a 'normal'
IPv6 network (i.e. the Internet).

> This requires the border router to have two interfaces, a downstream, 6lo-interface

Reviewer: ```suggestion
This requires the border router to have two interfaces, a downstream, 6LoWPAN-interface
```
(or `6Lo`, but that is not really a protocol)

> and an IPv6 uplink.

Reviewer: 6LoWPAN *is* IPv6 over X (where X is an LLN technology), so this wording is a bit confusing.

Author: I hope it's less confusing now.

> This setup comes with support for three uplink types pre-configured.

Reviewer: The three uplink types should be listed here, the default specified and `native` introduced as a special case
```suggestion
This example comes with support for three uplink types pre-configured:
- [`ethos`](https://doc.riot-os.org/group__drivers__ethos.html) (default)
- [`slip`](https://tools.ietf.org/html/rfc1055)
- `wifi`

For `native` the host-facing [`netdev_tap`](https://doc.riot-os.org/netdev__tap_8h.html) device
is configured, providing connectivity via a TAP interface to the RIOT instance.
 
```

> To select e.g. a SLIP uplink, add `UPLINK=slip` to your `make` command.

Reviewer: This wording puts unnecessary emphasis on SLIP. How about

> To select an uplink, set the UPLINK environment variable. For instance, use `UPLINK=slip` for a SLIP uplink.

`ethos` and `slip` will make use of the existing serial interface that is used for the
RIOT shell to provide an upstream interface. Your computer will act as the upstream
router, stdio is multiplexed over the same line.

The `wifi` uplink will connect to an existing WiFi (IEEE 802.11) network.

> The network must provide a DHCPv6 server that supports prefix delegation (IA_PD).

Reviewer: ```suggestion
The network must provide a DHCPv6 server that supports prefix delegation (IA_PD)
when `USE_DHCPV6=1` is set (default).
```

> Use `WIFI_SSID="SSID" WIFI_PASS="password"` in your `make` command to set your WiFi

Reviewer: ```suggestion
Use `WIFI_SSID="SSID" WIFI_PASS="password"` in your `make` command to set your WiFi's
```

credentials. You can alternatively edit the `Makefile`.

> Currently this requires an esp8266 or esp32 for the border router.

Reviewer: ```suggestion
Currently, `wifi` requires an esp8266 or esp32 for the border router.
```

> On the downstream interface the proprietary `esp_now` link is used. It behaves simmilar

Reviewer 2: ```suggestion
On the downstream interface the proprietary `esp_now` link is used. It behaves similar
```

> to IEEE 802.15.4 but is only availiably on Espressif's wireless MCUs.

Reviewer: Is the info about `esp_now` important? It seems more of a side note (as one can also use an external IEEE 802.15.4 radio as noted below). I rather would remove this as this is not related to the WiFi `UPLINK` configuration but rather how the user configures their downlink.

Author: Well `esp_now` is always available without the need to connect anything, so I think it's worth mentioning.
I've reworded it a bit though.

> to IEEE 802.15.4 but is only availiably on Espressif's wireless MCUs.

Reviewer: ```suggestion
to IEEE 802.15.4 but is only availaibly on Espressif's wireless MCUs.
```

to run 6LoWPAN on and an IPv6 uplink.

This example comes with support for three uplink types pre-configured:

 - [`ethos`](https://doc.riot-os.org/group__drivers__ethos.html) (default)
 - [`slip`](https://tools.ietf.org/html/rfc1055)
 - `wifi`

For `native` the host-facing [`netdev_tap`](https://doc.riot-os.org/netdev__tap_8h.html) device
is configured, providing connectivity via a TAP interface to the RIOT instance.
 
To select an uplink, set the UPLINK environment variable. For instance, use `UPLINK=slip`
for a SLIP uplink.

`ethos` and `slip` will make use of the existing serial interface that is used for the
RIOT shell to provide an upstream interface. Your computer will act as the upstream
router, stdio is multiplexed over the same line.

The `wifi` uplink will connect to an existing WiFi (IEEE 802.11) network.
The network must provide a DHCPv6 server that supports prefix delegation (IA_PD) when
`USE_DHCPV6=1` is set (default).

Use `WIFI_SSID="SSID" WIFI_PASS="password"` in your `make` command to set your WiFi's
credentials. You can alternatively edit the `Makefile`.

Currently, `wifi` requires an esp8266 or esp32 for the border router and will default

> to using `esp_now` for the downstream interface.

Reviewer: You seem insistent to keep `esp_now` in... Fine, I'm not fighting about minor details such as this...

Author: I mean it *is* the default, so that should be documented.

 - [`ethos`](https://doc.riot-os.org/group__drivers__ethos.html) (default)
 - [`slip`](https://tools.ietf.org/html/rfc1055)
 - `wifi`

For `native` the host-facing [`netdev_tap`](https://doc.riot-os.org/netdev__tap_8h.html) device
is configured, providing connectivity via a TAP interface to the RIOT instance.
 
To select an uplink, set the UPLINK environment variable. For instance, use `UPLINK=slip`
for a SLIP uplink.

`ethos` and `slip` will make use of the existing serial interface that is used for the
RIOT shell to provide an upstream interface. Your computer will act as the upstream
router, stdio is multiplexed over the same line.

The `wifi` uplink will connect to an existing WiFi (IEEE 802.11) network.
The network must provide a DHCPv6 server that supports prefix delegation (IA_PD) when
`USE_DHCPV6=1` is set (default).

Use `WIFI_SSID="SSID" WIFI_PASS="password"` in your `make` command to set your WiFi's
credentials. You can alternatively edit the `Makefile`.

Currently, `wifi` requires an esp8266 or esp32 for the border router.
On the downstream interface the proprietary `esp_now` link is used by default.

> It behaves similar to IEEE 802.15.4 but is only availaibly on Espressif's wireless MCUs.

Reviewer: ```suggestion
It behaves similar to IEEE 802.15.4 but is only available on Espressif's wireless MCUs.
```

You failed my test :P

```

## Usage
The `start_network.sh` script needed for `ethos` and `slip` is automatically run
if you type

```bash
make term
```

This will execute the needed commands to setup a `tap` (`ethos`) or `tun` (`slip`)
interface and configure the BR.
Notice that this will also configure `2001:db8::/64` as a prefix.
This prefix should be announced to other motes through the wireless interface.

> As said previously, `ethos` and `slip` allow to send IP packets and shell commands.

Reviewer: ```suggestion
As said previously, `ethos` and `slipdev` allow to send IP packets and shell commands.
```

As it is the device-specific implementation that allows for the multiplexing, not the SLIP spec.
